third-party/Makefile

```diff
@@ -88,31 +107,42 @@ else ifeq ($(wildcard $(GMP_H_FILE)),)
 endif
 
 gmp: $(GMP_H_FILE)
-$(GMP_H_FILE):
+$(GMP_H_FILE): $(GMP_DEPEND)
+	rm -rf $(GMP_INSTALL_DIR) && rm -rf $(GMP_BUILD_DIR)
 	cd gmp && $(MAKE)
 
 hwloc: $(HWLOC_INSTALL_DIR)
-$(HWLOC_INSTALL_DIR):
+$(HWLOC_INSTALL_DIR): $(HWLOC_DEPEND)
+	rm -rf $(HWLOC_INSTALL_DIR) && rm -rf $(HWLOC_BUILD_DIR)
 	cd hwloc && $(MAKE)
 
 jemalloc: $(JEMALLOC_INSTALL_DIR)
-$(JEMALLOC_INSTALL_DIR):
+$(JEMALLOC_INSTALL_DIR): $(JEMALLOC_DEPEND)
+	rm -rf $(JEMALLOC_INSTALL_DIR) && rm -rf $(JEMALLOC_BUILD_DIR)
 	cd jemalloc && $(MAKE)
 
 libunwind: $(LIBUNWIND_INSTALL_DIR)
-$(LIBUNWIND_INSTALL_DIR):
+$(LIBUNWIND_INSTALL_DIR): $(LIBUNWIND_DEPEND)
+	rm -rf $(LIBUNWIND_INSTALL_DIR) && rm -rf $(LIBUNWIND_BUILD_DIR)
 	cd libunwind && $(MAKE)
 
 massivethreads: $(MASSIVETHREADS_INSTALL_DIR)
-$(MASSIVETHREADS_INSTALL_DIR):
+$(MASSIVETHREADS_INSTALL_DIR): $(MASSIVETHREADS_DEPEND)
+	rm -rf $(MASSIVETHREADS_INSTALL_DIR) && rm -rf $(MASSIVETHREADS_BUILD_DIR)
 	cd massivethreads && $(MAKE)
 
+# Qthreads may use hwloc and/or jemalloc. Ensure they are built first
 ifeq ($(CHPL_MAKE_HWLOC), hwloc)
-qthread: $(HWLOC_INSTALL_DIR) $(QTHREAD_INSTALL_DIR)
-else
-qthread: $(QTHREAD_INSTALL_DIR)
+QTHREAD_DEPEND_INSTALL_DIRS += $(HWLOC_INSTALL_DIR)
 endif
-$(QTHREAD_INSTALL_DIR):
+ifeq ($(CHPL_MAKE_JEMALLOC), jemalloc)
+QTHREAD_DEPEND_INSTALL_DIRS += $(JEMALLOC_INSTALL_DIR)
+endif
+QTHREAD_DEPEND_INSTALL_DIRS += $(QTHREAD_INSTALL_DIR)
+qthread: $(QTHREAD_DEPEND_INSTALL_DIRS)
+
+$(QTHREAD_INSTALL_DIR): $(QTHREAD_DEPEND) $(HWLOC_DEPEND) $(JEMALLOC_DEPEND)
```

Comment:
Ok, do you need that to be part of this PR, or can that be a "if people notice a slowdown, fix it" sort of thing? My (relatively strong) preference would be to wait and see because of the impact making them conditional will have on the readability/maintainability of this code